File: flutter/android/src/main/kotlin/com/spruceid/sprucekit_mobile/Oid4vciAdapter.kt
package com.spruceid.sprucekit_mobile

import android.content.Context
import com.spruceid.mobile.sdk.KeyManager
import com.spruceid.mobile.sdk.Oid4vciAsyncHttpClient
import com.spruceid.mobile.sdk.rs.CredentialFormat as RsCredentialFormat
import com.spruceid.mobile.sdk.rs.CredentialOrConfigurationId
import com.spruceid.mobile.sdk.rs.CredentialResponse
import com.spruceid.mobile.sdk.rs.GrantType as RsGrantType
import com.spruceid.mobile.sdk.rs.InputMode as RsInputMode
import com.spruceid.mobile.sdk.rs.JwsSigner
import com.spruceid.mobile.sdk.rs.JwsSignerInfo
import com.spruceid.mobile.sdk.rs.Oid4vciCompatibilityMode
import com.spruceid.mobile.sdk.rs.Oid4vciException
import com.spruceid.mobile.sdk.rs.Oid4vciFacadeClient
import com.spruceid.mobile.sdk.rs.Oid4vciFacadeCredentialToken
import com.spruceid.mobile.sdk.rs.Oid4vciFacadeCredentialTokenState
import com.spruceid.mobile.sdk.rs.Oid4vciFacadeResolvedOffer
import com.spruceid.mobile.sdk.rs.Oid4vciFacadeWaitingForAuthorizationCode
import com.spruceid.mobile.sdk.rs.Proofs
import com.spruceid.mobile.sdk.rs.createJwtProof
import com.spruceid.mobile.sdk.rs.decodeDerSignature
import com.spruceid.mobile.sdk.rs.generateDidJwkUrl
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

/**
 * OID4VCI Pigeon Adapter
 *
 * Handles OpenID for Verifiable Credential Issuance flow
 */
internal class Oid4vciAdapter(private val context: Context) : Oid4vci {

    private data class SessionContext(
        val resolvedOffer: Oid4vciFacadeResolvedOffer,
        var tokenState: Oid4vciFacadeCredentialTokenState,
        val signer: JwsSigner,
        val clientId: String,
        val keyId: String,
        val httpClient: Oid4vciAsyncHttpClient,
        val redirectUrl: String?,
        var waitingForAuthCode: Oid4vciFacadeWaitingForAuthorizationCode? = null,
    )

    private val sessions = ConcurrentHashMap<String, SessionContext>()

    override fun runIssuance(
        credentialOffer: String,
        clientId: String,
        redirectUrl: String,
        keyId: String,
        didMethod: DidMethod,
        contextMap: Map<String, String>?,
        callback: (Result<Oid4vciResult>) -> Unit
    ) {
        CoroutineScope(Dispatchers.IO).launch {
            try {
                val result = performIssuance(
                    credentialOffer = credentialOffer,
                    keyId = keyId
                )
                callback(Result.success(result))
            } catch (e: Exception) {
                callback(Result.success(Oid4vciError(message = e.localizedMessage ?: "Unknown error")))
            }
        }
    }

    override fun parseOffer(
        credentialOffer: String,
        callback: (Result<ParsedOfferMetadata>) -> Unit
    ) {
        CoroutineScope(Dispatchers.IO).launch {
            try {
                val httpClient = Oid4vciAsyncHttpClient()
                val client = facadeClient("parse-offer-only")
                val offerUrl = normalizeOfferUrl(credentialOffer)
                val resolved = client.resolveOfferUrl(httpClient, offerUrl)
                callback(Result.success(buildParsedOfferMetadata(resolved)))
            } catch (e: Exception) {
                callback(Result.failure(e))
            }
        }
    }

    override fun acceptOffer(
        credentialOffer: String,
        clientId: String,
        keyId: String,
        didMethod: DidMethod,
        redirectUrl: String?,
        callback: (Result<OfferSession>) -> Unit
    ) {
        CoroutineScope(Dispatchers.IO).launch {
            try {
                val httpClient = Oid4vciAsyncHttpClient()
                val signer = buildJwsSigner(keyId)
                val derivedClientId = derivedClientId(keyId)
                val client = facadeClient(derivedClientId)
                val offerUrl = normalizeOfferUrl(credentialOffer)
                val resolved = client.resolveOfferUrl(httpClient, offerUrl)
                val tokenState = client.acceptOffer(httpClient, resolved)
                val sessionId = UUID.randomUUID().toString()
                sessions[sessionId] = SessionContext(
                    resolvedOffer = resolved,
                    tokenState = tokenState,
                    signer = signer,
                    clientId = derivedClientId,
                    keyId = keyId,
                    httpClient = httpClient,
                    redirectUrl = redirectUrl,
                )
                callback(Result.success(
                    OfferSession(
                        sessionId = sessionId,
                        metadata = buildParsedOfferMetadata(resolved),
                    )
                ))
            } catch (e: Exception) {
                callback(Result.failure(e))
            }
        }
    }

    override fun continueWithTxCode(
        sessionId: String,
        txCode: String,
        callback: (Result<Oid4vciResult>) -> Unit
    ) {
        CoroutineScope(Dispatchers.IO).launch {
            val ctx = sessions[sessionId]
            if (ctx == null) {
                callback(Result.success(Oid4vciError("session not found")))
                return@launch
            }
            val txState = ctx.tokenState as? Oid4vciFacadeCredentialTokenState.RequiresTxCode
            if (txState == null) {
                sessions.remove(sessionId)
                callback(Result.success(Oid4vciError("session not in tx_code state")))
                return@launch
            }
            try {
                val credentialToken = txState.v1.proceed(ctx.httpClient, txCode)
                val credentials = exchangeCredentialWithToken(ctx, credentialToken)
                sessions.remove(sessionId)
                callback(Result.success(Oid4vciSuccess(credentials)))
            } catch (e: Exception) {
                sessions.remove(sessionId)
                callback(Result.success(Oid4vciError(e.message ?: "unknown")))
            }
        }
    }

    override fun buildAuthorizationUrl(
        sessionId: String,
        callback: (Result<String?>) -> Unit
    ) {
        CoroutineScope(Dispatchers.IO).launch {
            val ctx = sessions[sessionId]
            if (ctx == null) {
                callback(Result.success(null))
                return@launch
            }
            val redirect = ctx.redirectUrl
            if (redirect.isNullOrEmpty()) {
                callback(Result.success(null))
                return@launch
            }
            val authState = ctx.tokenState as? Oid4vciFacadeCredentialTokenState.RequiresAuthorizationCode
            if (authState == null) {
                callback(Result.success(null))
                return@launch
            }
            try {
                val waiting = authState.v1.proceed(ctx.httpClient, redirect)
                ctx.waitingForAuthCode = waiting
                callback(Result.success(waiting.redirectUrl()))
            } catch (e: Exception) {
                callback(Result.success(null))
            }
        }
    }

    override fun continueWithAuthorizationCode(
        sessionId: String,
        code: String,
        callback: (Result<Oid4vciResult>) -> Unit
    ) {
        CoroutineScope(Dispatchers.IO).launch {
            val ctx = sessions[sessionId]
            if (ctx == null) {
                callback(Result.success(Oid4vciError("session not found")))
                return@launch
            }
            val waiting = ctx.waitingForAuthCode
            if (waiting == null) {
                sessions.remove(sessionId)
                callback(Result.success(Oid4vciError("session not awaiting authorization code")))
                return@launch
            }
            try {
                val credentialToken = waiting.proceed(ctx.httpClient, code)
                val credentials = exchangeCredentialWithToken(ctx, credentialToken)
                sessions.remove(sessionId)
                callback(Result.success(Oid4vciSuccess(credentials)))
            } catch (e: Exception) {
                sessions.remove(sessionId)
                callback(Result.success(Oid4vciError(e.message ?: "unknown")))
            }
        }
    }

    override fun releaseSession(
        sessionId: String,
        callback: (Result<Unit>) -> Unit
    ) {
        sessions.remove(sessionId)
        callback(Result.success(Unit))
    }

    // — Helpers —

    private fun normalizeOfferUrl(credentialOffer: String): String =
        if (credentialOffer.startsWith("openid-credential-offer://")) credentialOffer
        else "openid-credential-offer://$credentialOffer"

    /// Compatibility facade in Auto mode: prefers the OID4VCI 1.0 (final)
    /// request format and transparently retries with the legacy draft format
    /// when the issuer rejects the v1 credential request with a 400.
    private fun facadeClient(clientId: String): Oid4vciFacadeClient =
        Oid4vciFacadeClient(clientId, Oid4vciCompatibilityMode.AUTO)

    private fun derivedClientId(keyId: String): String {
        val keyManager = KeyManager()
        val jwk = keyManager.getOrInsertJwk(keyId)
        return generateDidJwkUrl(jwk).did().toString()
    }

    private fun buildJwsSigner(keyId: String): JwsSigner {
        val keyManager = KeyManager()
        val jwk = keyManager.getOrInsertJwk(keyId)
        val didUrl = generateDidJwkUrl(jwk)
        jwk.setKid(didUrl.toString())
        return object : JwsSigner {
            override suspend fun fetchInfo(): JwsSignerInfo = jwk.fetchInfo()
            override suspend fun signBytes(signingBytes: ByteArray): ByteArray =
                decodeDerSignature(keyManager.signPayload(keyId, signingBytes)
                    ?: throw IllegalStateException("signing failed for key: $keyId"))
        }
    }

    private fun buildParsedOfferMetadata(resolved: Oid4vciFacadeResolvedOffer): ParsedOfferMetadata {
        val grantType = when (resolved.grantType()) {
            RsGrantType.PRE_AUTH_CODE_NO_TX_CODE -> GrantType.PRE_AUTH_CODE_NO_TX_CODE
            RsGrantType.PRE_AUTH_CODE_WITH_TX_CODE -> GrantType.PRE_AUTH_CODE_WITH_TX_CODE
            RsGrantType.AUTHORIZATION_CODE -> GrantType.AUTHORIZATION_CODE
        }
        return ParsedOfferMetadata(
            issuerId = resolved.credentialIssuer(),
            issuerDisplayName = resolved.issuerDisplayName(),
            credentialConfigurationIds = resolved.credentialConfigurationIds(),
            grantType = grantType,
            txCode = buildTxCodeMetadata(resolved),
        )
    }

    private fun buildTxCodeMetadata(resolved: Oid4vciFacadeResolvedOffer): TxCodeMetadata? {
        val def = resolved.txCodeDefinition() ?: return null
        val inputMode = when (def.inputMode) {
            RsInputMode.TEXT -> TxCodeInputMode.TEXT
            RsInputMode.NUMERIC -> TxCodeInputMode.NUMERIC
        }
        return TxCodeMetadata(
            inputMode = inputMode,
            length = def.length?.toLong(),
            description = def.description,
        )
    }

    /// Exchange every credential in the offer against the token. Loops over
    /// the offer's credential_configuration_ids and issues a fresh nonce + JWT
    /// proof per credential (the protocol requires a unique proof per request).
    private suspend fun exchangeAllCredentials(
        httpClient: Oid4vciAsyncHttpClient,
        clientId: String,
        audience: String,
        signer: JwsSigner,
        token: Oid4vciFacadeCredentialToken,
        configIds: List<String>,
    ): List<IssuedCredential> {
        val result = mutableListOf<IssuedCredential>()
        for (configId in configIds) {
            val nonce = token.getNonce(httpClient)
            val jwt = createJwtProof(clientId, audience, null, nonce, signer)
            val proofs = Proofs.Jwt(listOf(jwt))
            val response = token.exchangeCredential(
                httpClient,
                CredentialOrConfigurationId.Configuration(configId),
                proofs,
            )
            when (response) {
                is CredentialResponse.Immediate -> {
                    result.addAll(response.v1.credentials.map { cred ->
                        val formatString = when (cred.format) {
                            is RsCredentialFormat.MsoMdoc -> "mso_mdoc"
                            is RsCredentialFormat.JwtVcJson -> "jwt_vc_json"
                            is RsCredentialFormat.JwtVcJsonLd -> "jwt_vc_json-ld"
                            is RsCredentialFormat.LdpVc -> "ldp_vc"
                            is RsCredentialFormat.Vcdm2SdJwt -> "vc+sd-jwt"
                            is RsCredentialFormat.DcSdJwt -> "dc+sd-jwt"
                            is RsCredentialFormat.Cwt -> "cwt"
                            is RsCredentialFormat.OpticalBarcodeCredential -> "optical_barcode_credential"
                            is RsCredentialFormat.Other -> (cred.format as RsCredentialFormat.Other).v1
                        }
                        IssuedCredential(payload = cred.payload.toString(Charsets.UTF_8), format = formatString)
                    })
                }
                is CredentialResponse.Deferred -> throw IllegalStateException("Deferred credentials not supported")
            }
        }
        return result
    }

    private suspend fun exchangeCredentialWithToken(
        ctx: SessionContext,
        token: Oid4vciFacadeCredentialToken,
    ): List<IssuedCredential> = exchangeAllCredentials(
        httpClient = ctx.httpClient,
        clientId = ctx.clientId,
        audience = ctx.resolvedOffer.credentialIssuer(),
        signer = ctx.signer,
        token = token,
        configIds = ctx.resolvedOffer.credentialConfigurationIds(),
    )

    private suspend fun performIssuance(
        credentialOffer: String,
        keyId: String
    ): Oid4vciResult {
        val httpClient = Oid4vciAsyncHttpClient()
        val signer = buildJwsSigner(keyId)
        val derivedClientId = derivedClientId(keyId)
        val oid4vciClient = facadeClient(derivedClientId)
        val offerUrl = normalizeOfferUrl(credentialOffer)
        val offer = oid4vciClient.resolveOfferUrl(httpClient, offerUrl)
        val credentialIssuer = offer.credentialIssuer()

        return when (val state = oid4vciClient.acceptOffer(httpClient, offer)) {
            is Oid4vciFacadeCredentialTokenState.RequiresAuthorizationCode -> {
                Oid4vciError(message = "Authorization Code Grant not supported")
            }
            is Oid4vciFacadeCredentialTokenState.RequiresTxCode -> {
                Oid4vciError(message = "Transaction Code not supported")
            }
            is Oid4vciFacadeCredentialTokenState.Ready -> {
                val credentialToken = state.v1
                val issuedCredentials = exchangeAllCredentials(
                    httpClient = httpClient,
                    clientId = derivedClientId,
                    audience = credentialIssuer,
                    signer = signer,
                    token = credentialToken,
                    configIds = offer.credentialConfigurationIds(),
                )
                Oid4vciSuccess(credentials = issuedCredentials)
            }
        }
    }
}

File: flutter/ios/Classes/Oid4vciAdapter.swift
import Foundation
import SpruceIDMobileSdk
import SpruceIDMobileSdkRs

/// JWS Signer implementation using KeyManager for OID4VCI proof of possession
class Oid4vciJwsSigner: JwsSigner, @unchecked Sendable {
    private let keyId: String
    private let jwk: Jwk

    init(keyId: String, jwk: Jwk) {
        self.keyId = keyId
        self.jwk = jwk
    }

    func fetchInfo() async throws -> JwsSignerInfo {
        return try await jwk.fetchInfo()
    }

    func signBytes(signingBytes: Data) async throws -> Data {
        guard let signature = KeyManager.signPayload(id: keyId, payload: [UInt8](signingBytes)) else {
            throw Oid4vciAdapterError.signingFailed
        }
        return try decodeDerSignature(signatureDer: Data(signature))
    }
}

/// Error types for OID4VCI adapter
enum Oid4vciAdapterError: Error {
    case signingFailed
}

// MARK: - Session Registry

private actor Oid4vciSessionRegistry {
    final class SessionContext {
        let resolvedOffer: ResolvedCredentialOffer
        var tokenState: CredentialTokenState
        let signer: Oid4vciJwsSigner
        let clientId: String
        let keyId: String
        let httpClient: Oid4vciAsyncHttpClient
        let redirectUrl: String?
        var waitingForAuthCode: WaitingForAuthorizationCode?

        init(
            resolvedOffer: ResolvedCredentialOffer,
            tokenState: CredentialTokenState,
            signer: Oid4vciJwsSigner,
            clientId: String,
            keyId: String,
            httpClient: Oid4vciAsyncHttpClient,
            redirectUrl: String?,
            waitingForAuthCode: WaitingForAuthorizationCode? = nil
        ) {
            self.resolvedOffer = resolvedOffer
            self.tokenState = tokenState
            self.signer = signer
            self.clientId = clientId
            self.keyId = keyId
            self.httpClient = httpClient
            self.redirectUrl = redirectUrl
            self.waitingForAuthCode = waitingForAuthCode
        }
    }

    private var sessions: [String: SessionContext] = [:]

    func insert(id: String, ctx: SessionContext) { sessions[id] = ctx }
    func get(id: String) -> SessionContext? { sessions[id] }
    func remove(id: String) { sessions.removeValue(forKey: id) }
}

// MARK: - Adapter

/// OID4VCI Pigeon Adapter for iOS
///
/// Handles OpenID for Verifiable Credential Issuance flow
class Oid4vciAdapter: Oid4vci {

    private let registry = Oid4vciSessionRegistry()

    // MARK: Protocol methods

    func parseOffer(
        credentialOffer: String,
        completion: @escaping (Result<ParsedOfferMetadata, Error>) -> Void
    ) {
        Task {
            do {
                let httpClient = Oid4vciAsyncHttpClient()
                let client = Oid4vciClient(clientId: "parse-offer-only")
                let offerUrl = normalizeOfferUrl(credentialOffer)
                let resolved = try await client.resolveOfferUrl(
                    httpClient: httpClient,
                    credentialOfferUrl: offerUrl
                )
                completion(.success(buildParsedOfferMetadata(resolved: resolved)))
            } catch {
                completion(.failure(error))
            }
        }
    }

    func runIssuance(
        credentialOffer: String,
        clientId: String,
        redirectUrl: String,
        keyId: String,
        didMethod: DidMethod,
        contextMap: [String: String]?,
        completion: @escaping (Result<Oid4vciResult, any Error>) -> Void
    ) {
        Task {
            do {
                let result = try await performIssuance(
                    credentialOffer: credentialOffer,
                    keyId: keyId
                )
                completion(.success(result))
            } catch {
                completion(.success(Oid4vciError(message: error.localizedDescription)))
            }
        }
    }

    func acceptOffer(
        credentialOffer: String,
        clientId: String,
        keyId: String,
        didMethod: DidMethod,
        redirectUrl: String?,
        completion: @escaping (Result<OfferSession, Error>) -> Void
    ) {
        Task {
            do {
                let httpClient = Oid4vciAsyncHttpClient()
                let signer = buildJwsSigner(keyId: keyId)
                let derivedClientId = derivedClientId(keyId: keyId)
                let oid4vciClient = Oid4vciClient(clientId: derivedClientId)
                let offerUrl = normalizeOfferUrl(credentialOffer)
                let resolved = try await oid4vciClient.resolveOfferUrl(
                    httpClient: httpClient,
                    credentialOfferUrl: offerUrl
                )
                let tokenState = try await oid4vciClient.acceptOffer(
                    httpClient: httpClient,
                    credentialOffer: resolved
                )
                let sessionId = UUID().uuidString
                let ctx = Oid4vciSessionRegistry.SessionContext(
                    resolvedOffer: resolved,
                    tokenState: tokenState,
                    signer: signer,
                    clientId: derivedClientId,
                    keyId: keyId,
                    httpClient: httpClient,
                    redirectUrl: redirectUrl
                )
                await registry.insert(id: sessionId, ctx: ctx)
                completion(.success(OfferSession(
                    sessionId: sessionId,
                    metadata: buildParsedOfferMetadata(resolved: resolved)
                )))
            } catch {
                completion(.failure(error))
            }
        }
    }

    func continueWithTxCode(
        sessionId: String,
        txCode: String,
        completion: @escaping (Result<Oid4vciResult, Error>) -> Void
    ) {
        Task {
            guard let ctx = await registry.get(id: sessionId) else {
                completion(.success(Oid4vciError(message: "session not found")))
                return
            }
            guard case .requiresTxCode(let txState) = ctx.tokenState else {
                await registry.remove(id: sessionId)
                completion(.success(Oid4vciError(message: "session not in tx_code state")))
                return
            }
            do {
                let token = try await txState.proceed(httpClient: ctx.httpClient, txCode: txCode)
                let credentials = try await exchangeCredentialWithToken(ctx: ctx, token: token)
                await registry.remove(id: sessionId)
                completion(.success(Oid4vciSuccess(credentials: credentials)))
            } catch {
                await registry.remove(id: sessionId)
                completion(.success(Oid4vciError(message: error.localizedDescription)))
            }
        }
    }

    func buildAuthorizationUrl(
        sessionId: String,
        completion: @escaping (Result<String?, Error>) -> Void
    ) {
        Task {
            guard let ctx = await registry.get(id: sessionId) else {
                completion(.success(nil))
                return
            }
            guard let redirect = ctx.redirectUrl, !redirect.isEmpty else {
                completion(.success(nil))
                return
            }
            guard case .requiresAuthorizationCode(let authState) = ctx.tokenState else {
                completion(.success(nil))
                return
            }
            do {
                let waiting = try await authState.proceed(httpClient: ctx.httpClient, redirectUrl: redirect)
                ctx.waitingForAuthCode = waiting
                completion(.success(waiting.redirectUrl()))
            } catch {
                completion(.success(nil))
            }
        }
    }

    func continueWithAuthorizationCode(
        sessionId: String,
        code: String,
        completion: @escaping (Result<Oid4vciResult, Error>) -> Void
    ) {
        Task {
            guard let ctx = await registry.get(id: sessionId) else {
                completion(.success(Oid4vciError(message: "session not found")))
                return
            }
            guard let waiting = ctx.waitingForAuthCode else {
                await registry.remove(id: sessionId)
                completion(.success(Oid4vciError(message: "session not awaiting authorization code")))
                return
            }
            do {
                let token = try await waiting.proceed(httpClient: ctx.httpClient, authorizationCode: code)
                let credentials = try await exchangeCredentialWithToken(ctx: ctx, token: token)
                await registry.remove(id: sessionId)
                completion(.success(Oid4vciSuccess(credentials: credentials)))
            } catch {
                await registry.remove(id: sessionId)
                completion(.success(Oid4vciError(message: error.localizedDescription)))
            }
        }
    }

    func releaseSession(
        sessionId: String,
        completion: @escaping (Result<Void, Error>) -> Void
    ) {
        Task {
            await registry.remove(id: sessionId)
            completion(.success(()))
        }
    }

    // MARK: - Helpers

    private func normalizeOfferUrl(_ credentialOffer: String) -> String {
        credentialOffer.starts(with: "openid-credential-offer://")
            ? credentialOffer
            : "openid-credential-offer://\(credentialOffer)"
    }

    private func buildJwsSigner(keyId: String) -> Oid4vciJwsSigner {
        let jwk = KeyManager.getOrInsertJwk(id: keyId)
        let didUrl = generateDidJwkUrl(jwk: jwk)
        jwk.setKid(kid: didUrl.description)
        return Oid4vciJwsSigner(keyId: keyId, jwk: jwk)
    }

    private func derivedClientId(keyId: String) -> String {
        let jwk = KeyManager.getOrInsertJwk(id: keyId)
        let didUrl = generateDidJwkUrl(jwk: jwk)
        return didUrl.did().description
    }

    private func buildParsedOfferMetadata(resolved: ResolvedCredentialOffer) -> ParsedOfferMetadata {
        let rsGrantType: SpruceIDMobileSdkRs.GrantType = resolved.grantType()
        let grantType: GrantType
        switch rsGrantType {
        case .preAuthCodeNoTxCode:
            grantType = .preAuthCodeNoTxCode
        case .preAuthCodeWithTxCode:
            grantType = .preAuthCodeWithTxCode
        case .authorizationCode:
            grantType = .authorizationCode
        }
        return ParsedOfferMetadata(
            issuerId: resolved.credentialIssuer(),
            issuerDisplayName: resolved.issuerDisplayName(),
            credentialConfigurationIds: resolved.credentialConfigurationIds(),
            grantType: grantType,
            txCode: buildTxCodeMetadata(resolved: resolved)
        )
    }

    private func buildTxCodeMetadata(resolved: ResolvedCredentialOffer) -> TxCodeMetadata? {
        guard let def = resolved.txCodeDefinition() else { return nil }
        let inputMode: TxCodeInputMode
        switch def.inputMode {
        case .numeric:
            inputMode = .numeric
        case .text:
            inputMode = .text
        }
        return TxCodeMetadata(
            inputMode: inputMode,
            length: def.length.map { Int64($0) },
            description: def.description
        )
    }

    /// Exchange every credential in the offer against the token. Loops over
    /// the offer's credential_configuration_ids and issues a fresh nonce + JWT
    /// proof per credential (the protocol requires a unique proof per request).
    private func exchangeAllCredentials(
        httpClient: Oid4vciAsyncHttpClient,
        clientId: String,
        audience: String,
        signer: Oid4vciJwsSigner,
        token: CredentialToken,
        configIds: [String]
    ) async throws -> [IssuedCredential] {
        let oid4vciClient = Oid4vciClient(clientId: clientId)
        var result: [IssuedCredential] = []
        for configId in configIds {
            let nonce = try await token.getNonce(httpClient: httpClient)
            let jwt = try await createJwtProof(
                issuer: clientId,
                audience: audience,
                expireInSecs: nil,
                nonce: nonce,
                signer: signer
            )
            let proofs = Proofs.jwt([jwt])
            let response = try await oid4vciClient.exchangeCredential(
                httpClient: httpClient,
                token: token,
                credential: .configuration(configId),
                proofs: proofs
            )
            switch response {
            case .deferred:
                throw NSError(
                    domain: "OID4VCI",
                    code: 1,
                    userInfo: [NSLocalizedDescriptionKey: "Deferred credentials not supported"]
                )
            case .immediate(let immediateResponse):
                result.append(contentsOf: immediateResponse.credentials.map { cred in
                    IssuedCredential(
                        payload: String(decoding: Data(cred.payload), as: UTF8.self),
                        format: formatToString(cred.format)
                    )
                })
            }
        }
        return result
    }

    private func exchangeCredentialWithToken(
        ctx: Oid4vciSessionRegistry.SessionContext,
        token: CredentialToken
    ) async throws -> [IssuedCredential] {
        return try await exchangeAllCredentials(
            httpClient: ctx.httpClient,
            clientId: ctx.clientId,
            audience: ctx.resolvedOffer.credentialIssuer(),
            signer: ctx.signer,
            token: token,
            configIds: ctx.resolvedOffer.credentialConfigurationIds()
        )
    }

    /// Convert native CredentialFormat enum to string for Pigeon API
    private func formatToString(_ format: SpruceIDMobileSdkRs.CredentialFormat) -> String {
        switch format {
        case .msoMdoc:
            return "mso_mdoc"
        case .jwtVcJson:
            return "jwt_vc_json"
        case .jwtVcJsonLd:
            return "jwt_vc_json-ld"
        case .ldpVc:
            return "ldp_vc"
        case .vcdm2SdJwt:
            return "vc+sd-jwt"
        case .dcSdJwt:
            return "dc+sd-jwt"
        case .cwt:
            return "cwt"
        case .opticalBarcodeCredential:
            return "optical_barcode_credential"
        case .other(let value):
            return value
        }
    }

    // MARK: - Legacy full-flow path (runIssuance wrapper)

    private func performIssuance(
        credentialOffer: String,
        keyId: String
    ) async throws -> Oid4vciResult {
        let httpClient = Oid4vciAsyncHttpClient()
        let signer = buildJwsSigner(keyId: keyId)
        let derivedClientId = derivedClientId(keyId: keyId)
        let oid4vciClient = Oid4vciClient(clientId: derivedClientId)
        let offerUrl = normalizeOfferUrl(credentialOffer)
        let offer = try await oid4vciClient.resolveOfferUrl(
            httpClient: httpClient,
            credentialOfferUrl: offerUrl
        )
        let credentialIssuer = offer.credentialIssuer()
        let state = try await oid4vciClient.acceptOffer(
            httpClient: httpClient,
            credentialOffer: offer
        )

        switch state {
        case .requiresAuthorizationCode(_):
            return Oid4vciError(message: "Authorization Code Grant not supported")
        case .requiresTxCode(_):
            return Oid4vciError(message: "Transaction Code not supported")
        case .ready(let credentialToken):
            let issuedCredentials = try await exchangeAllCredentials(
                httpClient: httpClient,
                clientId: derivedClientId,
                audience: credentialIssuer,
                signer: signer,
                token: credentialToken,
                configIds: offer.credentialConfigurationIds()
            )
            return Oid4vciSuccess(credentials: issuedCredentials)
        }
    }
}
